### PR TITLE
ONL-4183 Add keyboard navigation for dropdown search component - CTA and Search areas

### PR DESCRIPTION
**Pull request by jaimeperezpedra**
The keyboard navigation on the CTA and Search area will work with the tab key, and the items will work as it was from before nothing changes with the arrows keys

### CODE REVIEW

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 1]
```diff
@@ -312,9 +312,6 @@ export default {
       if (!this.isOpen) {
         // necessary to regain the focus after tab/enter keyboard event if search feature is active
         this.initialFocusedElement = this.$refs.popover.$el.querySelector(':focus');
-        if (this.isSearchEnabled) {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue]: beautifully this is covered by `@focus="isSearchInputFocus = true"` above

```
afterShow() -> focusSearch() -> this.$refs.searchInput.focus() -> @focus="isSearchInputFocus = true"
```

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 2]
```diff
@@ -325,7 +322,7 @@ export default {
     focusSearch() {
       this.$nextTick(() => {
         /* istanbul ignore else */
-        if (this.isOpen && this.$refs.searchInput) {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:313]: if `isSearchEnabled` then `this.$refs.searchInput` always exists, we should use the former for conditional logic

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 3]
```diff
@@ -367,32 +364,26 @@ export default {
       } else {
         nextItem = this.filteredItems.find(item => !item.disabled);
       }
-      if (this.hasCta() && this.isCtaAreaFocus) {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue]: moved below for separation of concerns, here we act on the `nextItem` after we do the `isCtaAreaFocus` blur

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 4]
```diff
@@ -367,32 +364,26 @@ export default {
...
+
+      if (this.hasCta() && this.isCtaAreaFocus) {
+        this.isCtaAreaFocus = false;
+      }
     },
     onTabKeyDown() {
-      let ctaAreaElementFocuseable;
-      if (this.hasCta()) {
-        ctaAreaElementFocuseable = this.$refs.ctaArea.querySelector(
+      if (this.isSearchEnabled && !this.isSearchInputFocus) {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:377]: I think this is more elegant

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 5]
```diff
@@ -344,7 +342,7 @@ export default {
       } else {
         // selecting an item might affect the position of the popover,
         // e.g. new item moves the trigger down
-        this.$refs.popover.update();
+        // this.$refs.popover.update();
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:345]: not super happy about having to comment this

File: src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { mount } from '@vue/test-utils';
+import { mount, shallowMount } from '@vue/test-utils';
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js:1]: could we achieve same thing without `shallowMount`? tried many stubs

**Review by matallo — COMMENTED** [review]
reworded the tests cases, reordered them, and added some more cases

tests seem to pass now, but had to comment out an `update`

**Review by matallo — APPROVED** [review]

**Review by jmesao — APPROVED** [review]

**Review by mcibique — CHANGES_REQUESTED** [review]

**Comment by mcibique** [comment]
3 bugs to solve:
1. If I open dropdown story and I hit tab multiple times without opening dropdown, I want to see focus jumping from one dropdown to next one.
2. If I tab to a dropdown, open it via space, and there is no CTA nor search, after hitting tab again, it should close it and jump to next dropdown. If there is CTA or search, after opening, first tab should jump to search, next tab to CTA, next tab should close the dropdown and jump to next dropdown
3. If I reload the frame in storybook and I open dropdown "Single value - with CTA/Loading/Search", the search is not focused when the dropdown is open. So if I hit the tab for the first time, it jumps to CTA first, then the next tab jumps to search. It should not behaving like this. It should go to search first, then to CTA.

**Comment by jaimeperezpedra** [comment]
@mcibique All bugs solved, thx for the help, pls check the last commit ty!

File: src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js [hunk 7]
```diff
@@ -235,30 +244,223 @@ describe('EcDropdownSearch - Keyboard navigation', () => {
...
+      expect(document.activeElement).toBe(document.body);
+      await openDropdown(wrapper);
+      expect(document.activeElement).toBe(document.body);
+
+      await wrapper.findByDataTest('ec-dropdown-search').trigger('keydown.tab');
+
+      expect(document.activeElement).toBe(wrapper.findByDataTest('cta-data-test').element);
+    });
+
+    it('should lost the focus and close if cta is enabled and is focused', async () => {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js:285]: typo here and through the document "lose"

File: src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js [hunk 8]
```diff
@@ -235,30 +244,223 @@ describe('EcDropdownSearch - Keyboard navigation', () => {
...
+      expect(document.activeElement).toBe(document.body);
+      await wrapper.findByDataTest('ec-dropdown-search').trigger('keydown.down');
+      expect(document.activeElement).toBe(document.body);
+
+      await wrapper.findByDataTest('ec-dropdown-search').trigger('keydown.tab');
+
+      expect(document.activeElement).toBe(wrapper.findByDataTest('cta-data-test').element);
+    });
+
+    it('should lost the focus and close if search is enabled', async () => {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js:329]: typo

File: src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js [hunk 9]
```diff
@@ -235,30 +244,223 @@ describe('EcDropdownSearch - Keyboard navigation', () => {
...
+        },
+        attachTo: elem,
+      });
+
+      expect(document.activeElement).toBe(document.body);
+      await wrapper.findByDataTest('ec-dropdown-search').trigger('keydown.tab');
+
+      expect(document.activeElement).toBe(document.body);
+    });
+    it('should not focus the cta if search and cta are enabled but the cta has nothing to focus', async () => {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js:443]: line break

Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.keyboard-navigation.spec.js:443]: "if cta is not focusable"

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 10]
```diff
@@ -352,9 +363,48 @@ export default {
       } else {
         nextItem = this.filteredItems.find(item => !item.disabled);
       }
+
       if (nextItem) {
         this.select(nextItem, { keyboardNavigation: true });
       }
+
+      this.lostFocus();
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:371]: "loseFocus"

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 11]
```diff
@@ -372,6 +422,7 @@ export default {
     closeViaKeyboardNavigation() {
       if (this.isOpen) {
         this.hide();
+        this.lostFocus();
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:425]: typo

File: src/components/ec-dropdown-search/ec-dropdown-search.vue [hunk 12]
```diff
@@ -405,6 +456,18 @@ export default {
         }
       }
     },
+    lostFocus() {
```
Comment (matallo) [inline src/components/ec-dropdown-search/ec-dropdown-search.vue:459]: typo

**Review by matallo — APPROVED** [review]
typos, the rest LGTM

**Review by mcibique — APPROVED** [review]

**Review by mgascam — APPROVED** [review]

**Pull request merged**